echion/threads.h

```diff
@@ -294,6 +303,72 @@ void ThreadInfo::unwind_tasks()
     }
 }
 
+// ----------------------------------------------------------------------------
+void ThreadInfo::unwind_greenlets(PyThreadState* tstate)
+{
+    std::unordered_set<GreenletInfo::ID> parent_greenlets;
+
+    // Collect all parent greenlets
+    std::transform(
+        greenlet_parent_map.cbegin(), greenlet_parent_map.cend(),
+        std::inserter(parent_greenlets, parent_greenlets.begin()),
+        [](const std::pair<GreenletInfo::ID, GreenletInfo::ID>& kv) { return kv.second; });
+
+    // Unwind the leaf greenlets
+    for (auto& greenlet_info : greenlet_info_map)
+    {
+        auto greenlet_id = greenlet_info.first;
+        auto& greenlet = greenlet_info.second;
+
+        if (parent_greenlets.find(greenlet_id) != parent_greenlets.end())
+            continue;
+
+        auto frame = PyList_GetItem(greenlet->frame_cell, 0);  // TODO: UNSAFE!
```

Comment:
Even though we're using C API without GIL here, it's safe to do so because we always set the 0th element in `greenlet_tracer()` function 
https://github.com/P403n1x87/echion/pull/108/files#diff-28e906bed45dae3a7c6291d1323fd7a9cd28deb7d4c9491dd3bfbad5f16d2321R74-R80
is that right?

Reply:
Ah well reminded! I have left this comment to make sure I'd come back to this. I think this operation is not safe, unless we acquire the `greenlet_info_map_lock`. However, the effect might be similar to acquiring the GIL, so it would be ideal if we can avoid it. To do so, we either implement support for lists, or drop the frame cell and simply update the `GreenletInfo` object from the greenlet tracer. It's only a question of where the use of a lock is less problematic in terms of performance.

Reply:
ah yeah, completely forgot about `greenlet_info_map_lock`. wouldn't current impl possibly lead to seg faults?

Reply:
Indeed. I cannot convince myself that this operation is safe, even though I haven't seen any segfaults originating from it yet. I have changed this so that we explicitly set the frame object, which we can handle already in a safe way.

Reply:
Great, thanks for clarification and updating the code.